### PR TITLE
fix(dot/network): check if peer supports protocol

### LINKED ISSUE #1571: dot/network: check if peer supports protocol before sending message
## Task summary

- currently we don't check if a peer supports a given protocol before attempting to open a stream with them/send a message over that stream
- this results in an error
- we should check before doing this that the peer supports the given protocol

### PR DESCRIPTION
**Pull request by EclesioMeloJunior**
## Changes

- Encapsulate `peerstore.SupportsProtocols` function
- Check if the peer supports protocol before handshake on `sendData` function

## Tests

```
go run go test ^Test_PeerSupportsProtocol$ github.com/ChainSafe/gossamer/dot/network
```

## Issues

- closes #1571

### CODE REVIEW

File: dot/network/notifications.go [hunk 1]
```diff
@@ -198,6 +198,11 @@ func (s *Service) createNotificationsMessageHandler(info *notificationsProtocol,
 }
 
 func (s *Service) sendData(peer peer.ID, hs Handshake, info *notificationsProtocol, msg NotificationsMessage) {
+	if support, err := s.host.supportsProtocol(peer, info.protocolID); err != nil || !support {
+		logger.Warn("the peer does not supports the protocol", "protocol", info.protocolID, "peer", peer, "err", err)
```
Comment (arijitAD) [inline dot/network/notifications.go:202]: This should be set at `debug` level as we may get this error often.

File: dot/network/host.go [hunk 2]
```diff
@@ -362,6 +362,17 @@ func (h *host) peers() []peer.ID {
 	return h.h.Network().Peers()
 }
 
+// getPeerProtocols returns all the supported protocols to one peer
```
Comment (arijitAD) [inline dot/network/host.go:365]: `getPeerProtocols` -> `supportsProtocol`

`supportsProtocol` doesn't return all the supported protocols.
`supportsProtocol` checks if `protocol` is supported by `peerID`.

**Review by arijitAD — APPROVED** [review]

File: dot/network/notifications.go (hunk 1 above)
Comment (EclesioMeloJunior) [reply dot/network/notifications.go]: Done!

File: dot/network/host.go (hunk 2 above)
Comment (EclesioMeloJunior) [reply dot/network/host.go]: Done!

File: dot/network/host.go [hunk 3]
```diff
@@ -362,6 +362,17 @@ func (h *host) peers() []peer.ID {
 	return h.h.Network().Peers()
 }
 
+// supportsProtocol checks if the protocol is supported by peerID
+// returns an error if could not get peer protocols
+func (h *host) supportsProtocol(peerID peer.ID, protocol protocol.ID) (bool, error) {
+	peerProtocols, err := h.h.Network().Peerstore().SupportsProtocols(peerID, string(protocol))
```
Comment (noot) [inline dot/network/host.go:368]: I think you can directly do `h.h.Peerstore().SupportsProtocols()`

**Review by noot — APPROVED** [review]
looks good!

**Review by timwu20 — APPROVED** [review]

**Pull request merged**